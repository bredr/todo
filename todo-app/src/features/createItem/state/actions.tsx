import { ActionType, createAction } from "typesafe-actions";

export const open = createAction("OPEN_CREATE");
export const close = createAction("CLOSE_CREATE");

export type Actions = ActionType<typeof open> | ActionType<typeof close>;

export default { open, close };
