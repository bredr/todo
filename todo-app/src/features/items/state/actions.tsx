import { ActionType, createAction } from "typesafe-actions";
import { grpcRequest } from "../../../middleware/grpc";
import { Items, Pagination } from "../../../proto/todo_pb";

export const setItems = createAction("SET_ITEMS", action => {
  return (items: Items.AsObject) => action(items);
});

export const loadingItems = createAction("LOADING_ITEMS");

export const getItems = grpcRequest<Pagination, Items>();

export type Actions =
  | ActionType<typeof setItems>
  | ActionType<typeof loadingItems>
  | ActionType<typeof getItems>;
