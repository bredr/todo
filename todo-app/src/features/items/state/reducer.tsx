import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Item } from "../../../proto/todo_pb";
import {
  Actions,
  cancelEditItem,
  editItem,
  getItems,
  loadingItems,
  setItems
} from "./actions";

interface IEditableItem extends Item.AsObject {
  edit: boolean;
}

export interface IState {
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
  readonly items: IEditableItem[];
  readonly loading: boolean;
}

const initState: IState = {
  items: [],
  limit: 100,
  loading: false,
  offset: 0,
  total: 0
};

export default createReducer<IState, Actions>(initState)
  .handleAction(loadingItems, (state, action) =>
    produce(state, draft => {
      draft.loading = true;
      draft.items = [];
      draft.limit = action.payload.limit;
      draft.offset = action.payload.offset;
    })
  )
  .handleAction(setItems, (state, action) =>
    produce(state, draft => {
      draft.loading = false;
      draft.limit = action.payload.limit;
      draft.offset = action.payload.offset;
      draft.total = action.payload.total;
      draft.items = action.payload.itemsList.map(item => ({
        ...item,
        edit: false
      }));
    })
  )
  .handleAction(editItem, (state, action) =>
    produce(state, draft => {
      draft.items = state.items.map(item =>
        item.id === action.payload ? { ...item, edit: true } : item
      );
    })
  )
  .handleAction(cancelEditItem, (state, action) =>
    produce(state, draft => {
      draft.items = state.items.map(item =>
        item.id === action.payload ? { ...item, edit: false } : item
      );
    })
  );
