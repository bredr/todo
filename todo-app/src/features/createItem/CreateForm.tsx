import { grpc } from "@improbable-eng/grpc-web";
import { createStyles, Grid, Theme } from "@material-ui/core";
import Button from "@material-ui/core/Button";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import TextField from "@material-ui/core/TextField";
import { makeStyles } from "@material-ui/styles";
import { Field, Form, Formik, FormikActions, FormikProps } from "formik";
import { Timestamp } from "google-protobuf/google/protobuf/timestamp_pb";
import React from "react";
import { useDispatch } from "react-redux";
import * as Yup from "yup";
import { Item } from "../../proto/todo_pb";
import { Todo } from "../../proto/todo_pb_service";
import { getItemsAction } from "../pagination/Pagination";
import * as toaster from "../toaster/state/actions";
import { CreateClose } from "./CreateClose";
import { FormikDatePicker } from "./DatePicker";
import actions from "./state/actions";

const DAY = 24 * 3.6e6;

interface IValues {
  description: string;
  daystocomplete: number;
  id?: string;
  due?: Date | 0;
}

interface IProps {
  item?: Item.AsObject;
  editing?: boolean;
}

const upsertItemAction = (values: IValues, id?: string) => {
  const request = new Item();
  const created = new Timestamp();
  created.setSeconds(Date.now());
  request.setCreated(created);
  request.setId(id || "");
  request.setDescription(values.description);
  request.setDaystocomplete(values.daystocomplete);
  if (values.due) {
    const due = new Timestamp();
    due.setSeconds(values.due.valueOf() / 1000);
    request.setDue(due);
  }
  return actions.upsertItem({
    methodDescriptor: Todo.UpsertItem,
    request,
    onMessage: () => getItemsAction(),
    onEnd: (code, message) =>
      code === grpc.Code.OK
        ? toaster.addSuccess("Updated")
        : toaster.addError(message)
  });
};

export const CreateForm: React.FC<IProps> = props => {
  const dispatch = useDispatch();
  const classes = useStyles();

  const initialValues = {
    description: (props.item && props.item.description) || "",
    daystocomplete: (props.item && props.item.daystocomplete) || 0,
    id: props.item && props.item.id,
    due:
      props.item &&
      props.item.due &&
      props.item.due.seconds &&
      new Date(props.item.due.seconds)
  };

  const onSubmit = (
    values: IValues,
    { setSubmitting }: FormikActions<IValues>
  ) => {
    const id = (props.item && props.item.id) || undefined;
    setSubmitting(true);
    dispatch(upsertItemAction(values, id));
    dispatch(actions.close());
    return;
  };
  return (
    <Card className={classes.card}>
      <CardContent>
        <Formik
          initialValues={initialValues}
          onSubmit={onSubmit}
          render={form}
          validationSchema={validationSchema}
          validateOnChange={true}
        />
      </CardContent>
    </Card>
  );
};

const useStyles = makeStyles({
  container: {
    display: "flex",
    flexWrap: "wrap"
  },
  textField: {
    width: 200
  },
  card: {
    minWidth: 275
  },
  bullet: {
    display: "inline-block",
    margin: "0 2px",
    transform: "scale(0.8)"
  },
  title: {
    fontSize: 14
  },
  pos: {
    marginBottom: 12
  }
});

const form = (props: FormikProps<IValues>) => {
  const {
    values: { description, daystocomplete, id },
    errors,
    touched,
    handleChange,
    isValid,
    setFieldTouched
  } = props;

  const change = (name: keyof IValues, e: React.ChangeEvent | any) => {
    e.persist();
    handleChange(e);
    setFieldTouched(name, true, false);
  };

  return (
    <Form>
      <Grid container={true} spacing={3}>
        <Grid item={true} xs={12}>
          <TextField
            id="description"
            name="description"
            helperText={touched.description ? errors.description : ""}
            error={touched.description && Boolean(errors.description)}
            label="Description"
            value={description}
            onChange={change.bind(null, "description")}
            fullWidth={true}
          />
        </Grid>
        <Grid item={true} xs={6}>
          <TextField
            id="daystocomplete"
            label="Days to complete"
            value={daystocomplete}
            onChange={change.bind(null, "daystocomplete")}
            helperText={touched.daystocomplete ? errors.daystocomplete : ""}
            error={touched.daystocomplete && Boolean(errors.daystocomplete)}
            type="number"
            InputLabelProps={{
              shrink: true
            }}
          />
        </Grid>
        <Grid item={true} xs={6}>
          <Field component={FormikDatePicker} name="due" />
        </Grid>
        <Grid item={true} xs={6}>
          <Button
            type="submit"
            variant="outlined"
            fullWidth={true}
            color="primary"
            disabled={!isValid}
          >
            Submit
          </Button>
        </Grid>
        <Grid item={true} xs={6}>
          <CreateClose {...{ id }} />
        </Grid>
      </Grid>
    </Form>
  );
};

export const validationSchema = Yup.object<IValues>().shape({
  description: Yup.string().required("Description is required"),
  daystocomplete: Yup.number()
    .required("Days to complete required")
    .min(0),
  due: Yup.date()
    .notRequired()
    .test("achievable", "Must be achievable", function(value) {
      const { daystocomplete } = this.parent;
      const currentDate = new Date(Date.now());
      const currentTime = currentDate.getHours();

      return (
        !value ||
        new Date(
          (currentTime > 10 ? DAY - currentTime : 0) +
            new Date(Date.now()).setHours(0, 0, 0, 0).valueOf() +
            daystocomplete * DAY
        ) <= value
      );
    })
});
