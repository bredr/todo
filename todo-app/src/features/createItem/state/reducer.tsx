import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Actions, close, open } from "./actions";

export interface IState {
  readonly isOpen: boolean;
}

const initState: IState = {
  isOpen: false
};

export default createReducer<IState, Actions>(initState)
  .handleAction(open, state =>
    produce(state, draft => {
      draft.isOpen = true;
    })
  )
  .handleAction(close, state =>
    produce(state, draft => {
      draft.isOpen = false;
    })
  );
