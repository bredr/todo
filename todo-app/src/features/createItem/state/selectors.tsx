import { createSelector } from "reselect";
import { IState } from "./reducer";

export const CREATE_SLICE = "CREATE_SLICE";

const sliceSelector = (state: { [CREATE_SLICE]: IState }) =>
  state[CREATE_SLICE];

const createItemOpen = createSelector(
  sliceSelector,
  state => state.isOpen
);

export default { createItemOpen };
