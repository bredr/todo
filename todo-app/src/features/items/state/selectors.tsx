import { createSelector } from "reselect";
import { IState } from "./reducer";

export const ITEMS_SLICE = "ITEMS_SLICE";

const sliceSelector = (state: { [ITEMS_SLICE]: IState }) => state[ITEMS_SLICE];

const items = createSelector(
  sliceSelector,
  state => state.items
);
const loading = createSelector(
  sliceSelector,
  state => state.loading
);
const hasNext = createSelector(
  sliceSelector,
  state => state.total > 0 && state.total - (state.limit + state.offset) > 0
);
const hasPrevious = createSelector(
  sliceSelector,
  state => state.offset > 0 && state.total > 0
);

export default { items, loading, hasNext, hasPrevious };
