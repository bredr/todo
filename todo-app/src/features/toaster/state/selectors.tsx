import { createSelector } from "reselect";
import { IState } from "./reducer";

export const TOASTER_SLICE = "TOASTER_SLICE";

const sliceSelector = (state: { [TOASTER_SLICE]: IState }) =>
  state[TOASTER_SLICE];

const currentMessage = createSelector(
  sliceSelector,
  state => (state.messages.length > 0 ? state.messages[0] : undefined)
);

export default { currentMessage };
