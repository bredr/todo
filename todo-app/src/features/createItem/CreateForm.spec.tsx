import { cleanup, fireEvent, render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { CreateForm, validationSchema } from "./CreateForm";
import { CREATE_SLICE } from "./state/selectors";

const mockStore = configureStore([]);

describe("CreateForm", () => {
  afterEach(() => cleanup());
  it("renders", () => {
    const store = mockStore({ [CREATE_SLICE]: { isOpen: true } });
    const { container } = render(
      <Provider store={store}>
        <CreateForm />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("validationSchema", () => {
    [
      { test: {}, valid: false },
      { test: { description: "test", daystocomplete: 1 }, valid: true },
      { test: { description: "test", daystocomplete: -1 }, valid: false },
      {
        test: {
          description: "test",
          daystocomplete: 1,
          due: new Date(Date.now() + 2 * 24 * 3.6e6)
        },
        valid: true
      },
      {
        test: {
          description: "test",
          daystocomplete: 1,
          due: new Date(Date.now())
        },
        valid: false
      }
    ].forEach((t, i) =>
      test(i.toString(), async () => {
        await expect(validationSchema.isValid(t.test)).resolves.toEqual(
          t.valid
        );
      })
    );
  });
});
