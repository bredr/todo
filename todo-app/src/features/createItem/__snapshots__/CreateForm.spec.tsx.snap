// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateForm renders 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-3 MuiPaper-rounded"
>
  <div
    class="MuiCardContent-root"
  >
    <form>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="description"
            >
              Description
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="description"
                name="description"
                type="text"
                value=""
              />
            </div>
            
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
              data-shrink="true"
              for="daystocomplete"
            >
              Days to complete
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="daystocomplete"
                type="number"
                value="0"
              />
            </div>
            
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
            format="MMMM do"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
              data-shrink="true"
            >
              due
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                readonly=""
                type="text"
                value="October 11th"
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Submit
            </span>
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Cancel
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
